### PR TITLE
CSHARP-3427: Can't make use of compression when application path contain space.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
EG: https://evergreen.mongodb.com/version/609d8e0e61837d3a55f5a870

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/NativeLibraryLoader/RelativeLibraryLocatorBase.cs [hunk 1]
```diff
@@ -26,24 +26,25 @@ internal abstract class RelativeLibraryLocatorBase : ILibraryLocator
         public virtual bool IsX32ModeSupported => false;
 
         // public methods
+        public virtual string GetBaseAssemblyUri() => typeof(RelativeLibraryLocatorBase).GetTypeInfo().Assembly.CodeBase;
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/NativeLibraryLoader/RelativeLibraryLocatorBase.cs:29]: the user's suggestion with assembly.Location doesn't work at least with cases when we configure `shadowCopy` in tests. See this [patch](https://evergreen.mongodb.com/version/609d37859ccd4e554a5c78b2) and .net framework variant.

File: src/MongoDB.Driver.Core/Core/NativeLibraryLoader/RelativeLibraryLocatorBase.cs [hunk 2]
```diff
@@ -26,24 +26,25 @@ internal abstract class RelativeLibraryLocatorBase : ILibraryLocator
...
             return GetAbsolutePath(relativePath);
         }
 
-        public virtual Assembly GetLibraryBaseAssembly()
+        public virtual string GetLibraryBaseAssemblyPath()
         {
-            return typeof(RelativeLibraryLocatorBase).GetTypeInfo().Assembly;
+            var baseAssemblyPathUri = GetBaseAssemblyUri();
+            var uri = new Uri(baseAssemblyPathUri);
+            return Uri.UnescapeDataString(uri.AbsolutePath);
```
Comment (JamesKovacs) [inline src/MongoDB.Driver.Core/Core/NativeLibraryLoader/RelativeLibraryLocatorBase.cs:41]: We are dealing with file paths, which are not URIs. So we shouldn't be using `Uri` to determine the absolute path. We should be using the `Path` API, which is intended for file paths.

```
public virtual string GetLibraryBaseAssemblyPath()
{
    var baseAssemblyPathUri = GetBaseAssemblyUri();
    return Path.GetFullPath(baseAssemblyPathUri);
}
```

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
We should be using `Path.GetFullPath()` not `Uri.AbsolutePath`. (See below.)

As well `RelativeLibraryLocatorBase` is an internal class, but has many public and virtual methods that don't need to be public or virtual. Similarly `GetBaseAssemblyUri` doesn't return a `Uri` but returns a path as a string. We should rename this method to `GetBaseAssemblyPath` or similar.

Overall we should remove our dependence on `Uri` entirely in this class and should instead use the file API (e.g. Path/File/Directory) instead.

File: src/MongoDB.Driver.Core/Core/NativeLibraryLoader/RelativeLibraryLocatorBase.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/NativeLibraryLoader/RelativeLibraryLocatorBase.cs:41]: the `baseAssemblyPathUri` is actually `Uri` in string form (with `file://` prefix), not sure how `Path.GetFullPath` works with this value and whether Uri-like value is expected input value, but i will check it

Comment (JamesKovacs) [reply src/MongoDB.Driver.Core/Core/NativeLibraryLoader/RelativeLibraryLocatorBase.cs:41]: It is strange how `Assembly.CodeBase` is marked obsolete and we are told to use `Assembly.Location` instead, but `Assembly.Location` says to use `Assembly.CodeBase` if you need the original and not shadow-copied path. So looks like we need to use `Assembly.CodeBase` even if it is marked obsolete.

Though I am curious why the shadow-copied assemblies can't load the unmanaged DLLs. I guess the unmanaged DLLs don't get shadow-copied along with their managed counterparts.

I did check and you are correct that `Assembly.CodeBase` is a URI as a string. Maybe `Uri.LocalPath` - which returns the local operating-system representation of a file name - will do the trick:

https://docs.microsoft.com/en-us/dotnet/api/system.uri.localpath?view=net-5.0

```
var uriPath = typeof(RelativeLibraryLocatorBase).GetTypeInfo().Assembly.CodeBase;
var uri = new Uri(uriPath);
var path = uri.LocalPath;
```

Comment (MikalaiMazurenka) [reply src/MongoDB.Driver.Core/Core/NativeLibraryLoader/RelativeLibraryLocatorBase.cs:41]: `Assembly.CodeBase` in fact contains a path without escaped characters and only differs in slashes and `file:///` prefix. And by using `Uri` you escape/unescape the characters, which isn't guaranteed to work properly. Consider the example where this wouldn't work and consider using the third option in the test I provided: https://gist.github.com/MikalaiMazurenka/b5fff9af75ebeb0c1cf645c6793519d4

File: tests/MongoDB.Driver.Core.Tests/Core/NativeLibraryLoader/NativeLibraryLoaderTests.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+using System.IO;
+using FluentAssertions;
+using MongoDB.Driver.Core.NativeLibraryLoader;
+using MongoDB.Driver.TestHelpers;
+using MongoDB.Shared;
+using Xunit;
+
+namespace MongoDB.Driver.Core.Tests.Core.NativeLibraryLoader
+{
+    public class NativeLibraryLoaderTests
```
Comment (MikalaiMazurenka) [inline tests/MongoDB.Driver.Core.Tests/Core/NativeLibraryLoader/NativeLibraryLoaderTests.cs:26]: It was hard to figure out what's going on here. Consider providing additional comments.

**Review by MikalaiMazurenka — COMMENTED** [review]
Great work! I'm ready to approve it as it, but I would like you to consider some suggestions first.

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Comment by DmitryLukyanov** [comment]
@MikalaiMazurenka found a new issue in this logic that is related to handling `#` sign. Uri has a special handling of this where it places all content after `#` (including `#`) into a special field called `Fragment`, so the solution can be just:
>return Uri.UnescapeDataString(uri.AbsolutePath + uri.Fragment)

where Fragment is a string empty if there is no `#`: https://docs.microsoft.com/en-us/dotnet/api/system.uri.fragment?view=net-5.0#remarks 
The good part of the original solution was that there were no changes in logic itself other than fixing in escaping logic, but if we want to fix the new issue also, we don't have a choice.

Regarding the suggestion with just manual `Replace` everything we don't need to, I would avoid it because:
* We have "strange" but at the same time built-in logic for Uri handling so I would prefer it.
* The manual changes here look like a platform-dependent or at least can be platform-dependent in future.

So, @MikalaiMazurenka , @JamesKovacs please take a look at this comment. 
If we decide to fix a new issue now, I created a temp branch with suggested solution [here](https://github.com/DmitryLukyanov/mongo-csharp-driver/tree/csharp3427_suggestion), also related [EG](https://evergreen.mongodb.com/version/60b11741d1fe0727fb1f79ef).

**Comment by DmitryLukyanov** [comment]
note: the mentioned exception https://docs.microsoft.com/en-us/dotnet/api/system.uri.fragment?view=net-5.0#exceptions is the same as here https://docs.microsoft.com/en-us/dotnet/api/system.uri.absolutepath?view=net-5.0#exceptions (also checked it in the source)

File: src/MongoDB.Driver.Core/Core/NativeLibraryLoader/RelativeLibraryLocatorBase.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/NativeLibraryLoader/RelativeLibraryLocatorBase.cs:41]: >you escape/unescape the characters, which isn't guaranteed to work properly

I'm not sure what do you mean (other than handling of `#`), from my understanding escaping/unescaping should always work in both ways with the same results

File: tests/MongoDB.Driver.Core.Tests/Core/NativeLibraryLoader/NativeLibraryLoaderTests.cs (hunk 3 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/NativeLibraryLoader/NativeLibraryLoaderTests.cs:26]: will do

**Comment by JamesKovacs** [comment]
It feels like we are plugging holes rather than solving the actual problem. What we need to know is the location where we are loading files from irregardless of shadow copying. Since shadow copying is controlled by the AppDomain, what does `AppDomain.CurrentDomain.BaseDirectory` point to when shadow copying? Is it the original assembly location prior to shadow copying? And if that’s the case, does it work on all our supported TFMs?

**Comment by DmitryLukyanov** [comment]
Discussed offline, the current next step is creating a new ticket that will consider additional (not space related) issues and possible refactoring of this logic

File: tests/MongoDB.Driver.Core.Tests/Core/NativeLibraryLoader/NativeLibraryLoaderTests.cs (hunk 3 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/NativeLibraryLoader/NativeLibraryLoaderTests.cs:26]: done

**Comment by DmitryLukyanov** [comment]
follow-up ticket: https://jira.mongodb.org/browse/CSHARP-3693

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Review by MikalaiMazurenka — APPROVED** [review]
LGTM

**Pull request merged**